src/messageforwarder.py

```diff
@@ -0,0 +1,241 @@
+        self._ban_appeal_channel = ban_appeal_channel
+        self._home_server = home_server
+
+        # Maps user ids to reply thread ids - used when receiving a DM to know which thread to forward it to (if create_threads is true)
+        self._user_id_to_thread_id = LRUCache(lambda user_id: db.get_user_reply_thread_id(user_id), maxsize=50)
+
+        # Maps user reply thread ids to user ids - used when staff replies in a thread to know which user to send the reply to (this will work for existing threads even if create_threads is false)
+        self._thread_id_to_user_id = LRUCache(lambda thread_id: db.get_user_reply_thread_user_id(thread_id), maxsize=50)
+
+        # Both of the above use an LRU cache wrapper around accessing the DB so that every DM/reply does not trigger DB access
+        # The chosen cache size should be equal to the number of concurrent/active conversations we expect users to have with bouncer
+
+    async def on_dm(self, message: discord.Message):
+        """
+        On a DM, forward the message to staff.
+
+        :param message: The message that was sent.
+        """
+        # Ignore blocked users
+        if commands.bu.is_in_blocklist(message.author.id):
+            return
+
+        # Figure out what message to send, which channel to send in, etc based on which server the sender is in
+        if is_in_home_server(message.author):  # If the user is in the home server, treat it as a regular DM
+            reply_message = f"<@{message.author.id}>"
+            answering_machine = commands.reply_am
+            reply_channel = client.get_channel(self._mailbox_channel)
+        else:  # Otherwise, assume it's a ban appeal (users must have a mutual server to message bouncer, they should only be able to join those two)
+            reply_message = f"{str(message.author)} ({message.author.id})"  # Can't ping user, so show username details
+            answering_machine = commands.ban_am
+            reply_channel = client.get_channel(self._ban_appeal_channel)
+
+        # Set latest received reply so '^' works when sending the reply command
+        answering_machine.set_recent_reply(message.author)
+
+        # Fill in the rest of the message with what the user said
+        content = commonbot.utils.combine_message(message)
+        reply_message += f": {content}"
+
+        # If forwarded messages should be grouped into threads, get or create the appropriate thread for the message user
+        if self._create_threads:
+            reply_channel = await self._get_or_create_user_reply_thread(message.author, reply_channel)
+
+        # Forward the message to the channel/thread
+        log_mes = await commonbot.utils.send_message(reply_message, reply_channel)
+
+        try:
+            # Send staff a message so they know something actually happened
+            await message.channel.send("Your message has been forwarded!")
+        except discord.errors.Forbidden as err:
+            if err.code == 50007:
```

Comment:
Optional: While we're changing this, we may want to double check if this is right/the best way to do this. I remember this being a bit of a slapped on solution.

Reply:
The exception/error code check?

Looking at all the exception types (https://discordpy.readthedocs.io/en/stable/api.html#exception-hierarchy) there doesn't look to be a better one.

Looking at discord's error code reference, they do document that one, so it should be stable: https://discord.com/developers/docs/topics/opcodes-and-status-codes#json

A Github search seems to show other people use that error code too: https://github.com/search?q=50007+discord&type=code

So I'm not sure how to do it better. We probably want to log other error codes though, I'll add a case for that.

Also I see I made a typo, that should say "Send the user", not "Send staff", I'll fix that.